<a href="https://colab.research.google.com/github/JerryYu0815/Jie_Yu_CSC_448_AI/blob/main/CSC448_Assignment_3_Jie_Yu.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

\(\^Be sure to update this button to point to your notebook instead of the sample notebook\)

In [10]:
# Imports section
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn import linear_model
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [11]:
# Using pandas load the dataset (load remotely, not locally)
df = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
df.head()

# Output the first 15 rows of the data
print("First 15 rows of the data:\n",df.iloc[0:15])

# Display a summary of the table information (number of datapoints, etc.)
print("\nTable information:")

#df['Size nm^3']= df['Size nm^3'].round(5).astype('int')
df.dtypes

First 15 rows of the data:
     Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Table information:


Temperature °C      int64
Mols KCL            int64
Size nm^3         float64
dtype: object

## Part 2. Splitting the dataset

In [12]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df[["Temperature °C","Mols KCL"]]  #input are temperatures and mols KCL
X.head()

Temperature °C  Mols KCL
0             469       647
1             403       694
2             302       975
3             779       916
4             901        18

In [13]:
y = df[["Size nm^3"]]
y.head()

Size nm^3
0  6.244743e+05
1  5.779610e+05
2  6.196847e+05
3  1.460449e+06
4  4.325726e+04

In [14]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

X_train, X_test, y_train, y_test = train_test_split(X.values, y.values, test_size=0.1, random_state=0)
print("Length of each set is:")
print('X_train: ',len(X_train))
print('X_test :',len(X_test))
print('y_train: ',len(y_train))
print('y_test: ',len(y_test))


Length of each set is:
X_train:  900
X_test : 100
y_train:  900
y_test:  100


## Part 3. Perform a Linear Regression

In [15]:
# Use sklearn to train a model on the training set
regr = linear_model.LinearRegression()
regr.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
pred = regr.predict(np.array([X_test[0]]))

print("The prediction result for X_test[0]= ",pred)
print("The actual reuslt y_test[0]= ",y_test[0])
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = regr.score(X_train, y_train)
print('score = ', score) 

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coef = regr.coef_
intercept = regr.intercept_
print('coef = ', coef)
print('intercept= ', intercept)

The prediction result for X_test[0]=  [[134891.57109752]]
The actual reuslt y_test[0]=  [170568.4286]
score =  0.8571870071495865
coef =  [[ 863.58108791 1006.12741921]]
intercept=  [-400305.91333353]


## The prediction result for X_test[0] = 134891.57109751 and the The actual reuslt y_test[0] = 170568.4286, which means that the prediction is not accurate.

## The score gotten in part 3 foris 0.85719 which is not so good. A scroe of 1 means that it is a perfect fit and the model is reliable while a score of 0 means that the calculation failed in modeling the data.

## Equation: $ y = 863.58109x_1 + 1006.12742x_2 - 400305.91333 $



## Part 4. Use Cross Validation

In [16]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(regr,X,y,cv = 100)
print("mean = ", (scores.mean()))
print("standar deviation = ", (scores.std()))
print("scores from 100 shuffles of the data\n",scores)


mean =  0.8088156665899868
standar deviation =  0.14982832048114586
scores from 100 shuffles of the data
 [0.72987142 0.02332136 0.64663948 0.83710031 0.81287812 0.85643292
 0.86566207 0.82943342 0.85717261 0.85818757 0.88407446 0.21899099
 0.79353519 0.93228937 0.74819809 0.91932985 0.85291091 0.72308356
 0.64174485 0.89081781 0.70643469 0.86544004 0.8450951  0.70336182
 0.86728463 0.91760202 0.90908958 0.94150897 0.83819493 0.87273024
 0.7871197  0.88467529 0.86671862 0.8176053  0.86637483 0.9025546
 0.85450024 0.88911985 0.81337559 0.62209156 0.88511368 0.87383856
 0.93874597 0.87945152 0.86422446 0.85185362 0.80632506 0.85400314
 0.90590826 0.85225702 0.65832    0.87102599 0.71313585 0.89400345
 0.58732207 0.87049636 0.88042279 0.88289086 0.34984312 0.86334775
 0.89350964 0.884054   0.83945642 0.91647922 0.84972854 0.88619777
 0.85252309 0.87032873 0.77491956 0.91480937 0.6467444  0.957948
 0.85213029 0.67112896 0.88361835 0.83713072 0.87391655 0.67181693
 0.83559998 0.92893673 0.8

## Part 5. Using Polynomial Regression

In [17]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(2)
X_train = poly.fit_transform(X_train)
X_test = poly.fit_transform(X_test)
y_test = np.array(y_test)

model = linear_model.LinearRegression()
model.fit(X_train,y_train)
print("The score is: ",model.score(X_train,y_train))
print("The coef is : ",model.coef_)
print("The intercept is : ",model.intercept_)
print("The first row of values in X_test = ",X_test[0])
print("The first row of values in y_test = ",y_test[0])
print("Predit the result for X_test[0]: ", model.predict([X_test[0]]))



# Report on the metrics and output the resultant equation as you did in Part 3.

The score is:  1.0
The coef is :  [[ 0.00000000e+00  1.20000000e+01 -1.10385822e-07 -1.20952137e-11
   2.00000000e+00  2.85714287e-02]]
The intercept is :  [1.39031908e-05]
The first row of values in X_test =  [1.0000e+00 3.1100e+02 2.6500e+02 9.6721e+04 8.2415e+04 7.0225e+04]
The first row of values in y_test =  [170568.4286]
Predit the result for X_test[0]:  [[170568.42856638]]


## The prediction result for X_test[0] = 170568.42856638 and the The actual reuslt y_test[0] = 170568.4286, which means that the prediction is very accurate.

## The score gotten in part 5 foris 1.0 which mean that is a perfect fit model.

## Equation: $ y = 0 + 12x_1 - 0*x_2 + 0*x_1^2+2x_1x_2+0.02857x_2^2 = 12x_1+2x_1x_2+0.02857x_2^2$

## Test equation for the polynomial regression model

In [18]:
x1 = 311
x2 = 265
y = 12*x1 + 2*x1*x2+0.02857*(x2 ** 2)
print("Expected: y = 170568.4286")
print("Equation: y = ", y)


Expected: y = 170568.4286
Equation: y =  170568.32825


## The equation is correct for our polynomial regression models.